Digitize focussed DS proposal (#1158)
IBM/project-ai-services: docs/proposals/data-source-connectors/digitize-connector-processing-proposal.md

> **Document metadata:** When a document row is created the scanner also writes the fingerprint into `documents.metadata` as `metadata.source_checksum`. This makes the content fingerprint visible in the document record without a registry join.
>
> ```json
> {
>   "source_checksum": "0234031ed6cb7d686152f45c38f41bc6-13",
>   "source_type": "s3",
>   "bucket": "ai-services",
>   "key": "reports/sg248590-2.pdf"
> }
> ```

**Checksum format reference:**

| Source | Value stored in `sha256` | Dedup property |
| --- | --- | --- |
| S3 single-part | S3 ETag = `MD5(file_bytes)` — 32-char hex, no suffix | Unique per file content (for this upload method) |
| S3 multi-part | S3 ETag = `MD5(MD5(p₁)‖…‖MD5(pₙ))-N` — hex + `-N` suffix | Unique per (file content + part size) |
| SFTP | `md5sum` output from remote host — 32-char hex | Unique per file content, protocol-independent |

### 4.4 `connector_file_membership`

Maps connectors to the `sha256` values they currently own.

```sql
CREATE TABLE IF NOT EXISTS connector_file_membership (

>     connector_id  TEXT NOT NULL,

Reviewer: this field should be generic to accommodate the user submitted jobs

Author: Added details for user submitted in latest commit

# Digitize — Data Source Connector Processing Proposal

> **Scope:** Internal `digitize` behavior after catalog sends connector payloads. Catalog-side concerns such as key management, connector CRUD, deployment wiring and TLS provisioning remain out of scope and are treated as infrastructure-level prerequisites.

---

## 1. Preconditions

Before any `digitize` connector endpoint is called:

- Catalog has already validated the remote connector configuration.
- Catalog sends secret material in plaintext via API calls:
  - `ssh`: `private_key`
  - `s3`: `secret_access_key`
- `digitize` encrypts those secret fields at rest using `/run/secrets/connector_encryption_key` before persisting them.
- `/run/secrets/connector_api_token` and `/run/secrets/connector_encryption_key` are mounted before pod start.
- The `file_checksum_registry` table and `FileChecksumRegistry` ORM model are already implemented (user-submitted documents only). Connector code must never read from or write to it — connector dedup is handled exclusively via `connector_file_membership` (see §4).

---

## 2. System Overview

### 2.1 Architecture Diagram

> ![Architecture Diagram](architecture-diagram.svg)

Reviewer: Arch diagram need some rewording 
SHA256 to md5
f.c.r table need to be removed
c.f.m table need to be renamed

Author: Done

### 2.2 Runtime Flow

```text
── Attach (POST /v1/connectors) ─────────────────────────────────────
  → validate + encrypt credentials

>   → INSERT active_connectors row

Reviewer: why did we start calling `active_connectors`?
should it be just connectors similar to jobs, documents..

Author: Done

  → schedule worker startup (background task)
  → worker thread starts and runs first tick immediately

── Config update (PUT /v1/connectors/{id}) ──────────────────────────
  → merge + re-encrypt changed fields
  → UPDATE active_connectors row
  → running worker re-reads config from DB before its next tick

── Each sync tick (connector-sourced) ───────────────────────────────
  Step 1 — load state

>     → SELECT known_checksums FROM connector_file_membership

Reviewer: how about?
connector_file_membership - connector_document_checksum
file_checksum_registry - document_checksum

Author: Done

# Digitize — Data Source Connector Processing Proposal

> **Scope:** Internal `digitize` behavior after catalog sends connector payloads. Catalog-side concerns such as key management, connector CRUD, deployment wiring and TLS provisioning remain out of scope and are treated as infrastructure-level prerequisites.

---

## 1. Preconditions

Before any `digitize` connector endpoint is called:

- Catalog has already validated the remote connector configuration.
- Catalog sends secret material in plaintext via API calls:
  - `ssh`: `private_key`
  - `s3`: `secret_access_key`
- `digitize` encrypts those secret fields at rest using `/run/secrets/connector_encryption_key` before persisting them.

> - `/run/secrets/connector_api_token` and `/run/secrets/connector_encryption_key` are mounted before pod start.

Reviewer: remove connector_api_token

Author: Done

`connection_details` for `ssh`:

| Field | Type | Required | Notes |
| --- | --- | --- | --- |
| `host` | `string` | ✅ | SFTP hostname |
| `username` | `string` | ✅ | |
| `remote_path` | `string` | ✅ | |
| `private_key` | `string` | ✅ | |

`connection_details` for `s3`:

| Field | Type | Required | Notes |
| --- | --- | --- | --- |
| `host` | `string` | ✅ | S3 endpoint hostname (e.g. `s3.us-east-1.amazonaws.com`) |
| `endpoint_url` | `string` | ✅ | Full S3 endpoint URL. AWS S3: `https://s3.<region>.amazonaws.com`. IBM COS: `https://s3.<region>.cloud-object-storage.appdomain.cloud`. Provider and region are auto-detected from this URL — no separate `region` field needed. |
| `bucket_name` | `string` | ✅ | |
| `access_key_id` | `string` | ✅ | IAM key ID (AWS) or HMAC key ID (IBM COS) |
| `secret_access_key` | `string` | ✅ | IAM secret (AWS) or HMAC secret (IBM COS) |
| `prefix` | `string` | ❌ | Key prefix to scope listing — empty means bucket root |
| `delimiter` | `string` | ❌ | Set `"/"` for non-recursive (immediate children only) |

> > **Checksum-based dedup:** For S3 connectors, `list_objects_v2` returns the object ETag at no extra API cost — stored as `checksum` in `connector_document_checksum`. If the checksum is already present for this connector the file is **never downloaded**. The checksum is also stored in `documents.metadata.source_checksum` for traceability.

Reviewer: ```suggestion
> **Checksum-based dedup:** For S3 connectors, `list_objects_v2` returns the object ETag at no extra API cost — stored as `checksum` in `connector_document_checksum`. If the checksum is already present for this connector the file is **never downloaded**.
```

Reviewer: Hope the removed line is not valid

Author: Yes, just stale reference, removed

#### Example payloads

```json
{
  "connector_id": "c7f3a2d1-...",
  "connector_name": "prod-sftp-reports",
  "type": "ssh",
  "allowed_extensions": [".pdf", ".docx"],
  "connection_details": {
    "host": "sftp.example.com",
    "username": "sync_user",
    "remote_path": "/exports/reports",
    "private_key": "-----BEGIN OPENSSH PRIVATE KEY-----\n..."
  }
}
```

```json
{
  "connector_id": "a1b2c3d4-...",
  "connector_name": "prod-s3-rag-docs",
  "type": "s3",
  "allowed_extensions": [".pdf", ".docx"],
  "connection_details": {
    "host": "s3.us-east-1.amazonaws.com",
    "endpoint_url": "https://s3.us-east-1.amazonaws.com",
    "bucket_name": "my-rag-documents",
    "prefix": "reports/",
    "delimiter": "/",
    "access_key_id": "AKIAIOSFODNN7EXAMPLE",
    "secret_access_key": "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
  }
}
```

IBM COS example:

```json
{
  "connector_id": "b5c6d7e8-...",
  "connector_name": "prod-cos-ai-services",
  "type": "s3",
  "allowed_extensions": [".pdf", ".docx"],
  "connection_details": {
    "host": "s3.us.cloud-object-storage.appdomain.cloud",

>     "endpoint_url": "https://s3.us.cloud-object-storage.appdomain.cloud",

Reviewer: why do we need two properties?
let's keep only endpoint_url

Author: Done

Response codes:

| Status | Meaning |
| --- | --- |
| `204 No Content` | Connector removed |
| `404 Not Found` | Connector does not exist |
| `409 Conflict` | A sync tick is currently running; retry after the tick completes |

> **Future:** when job cancellation is introduced, the `409` guard will be replaced by an interrupt call so that DELETE can succeed at any stage of the lifecycle.

### 3.4 `GET /v1/connectors`

Lists active connectors with non-secret configuration and current sync state.

Returned fields include:

- connector identity and config
- `sync_status`, `last_sync_at`, `last_sync_error`, `attached_at`, `total_files`

#### Example response

```json
[
  {
    "connector_id": "c7f3a2d1-4e5b-4c6d-8f9a-0b1c2d3e4f5a",
    "connector_name": "prod-sftp-reports",
    "type": "ssh",
    "attached_at": "2025-01-10T08:00:00Z",
    "last_sync_at": "2025-01-15T14:32:10Z",
    "sync_status": "up to date",
    "last_sync_error": null,
    "total_files": 42
  },
  {
    "connector_id": "a1b2c3d4-5e6f-7a8b-9c0d-1e2f3a4b5c6d",
    "connector_name": "prod-s3-rag-docs",
    "type": "s3",
    "attached_at": "2025-01-12T09:15:00Z",
    "last_sync_at": "2025-01-15T14:30:00Z",

>     "sync_status": "failed",

Reviewer: can we use out of sync instead of failed?

Author: Done

  "sync_interval_seconds": 300,
  "attached_at": "2025-01-10T08:00:00Z",
  "last_sync_at": "2025-01-15T14:32:10Z",
  "sync_status": "up to date",
  "connection_details": {
    "host": "sftp.example.com",
    "username": "sync_user",
    "remote_path": "/exports/reports"
  },
  "total_files": 42,
  "new_files": 2,
  "removed_files": 0,
  "failed_files": 2
}
```

### 3.6 `GET /v1/connectors/{connector_id}/sync-history`

Returns paginated tick history.

Query params:

| Param | Default | Notes |
| --- | --- | --- |
| `limit` | `50` | capped at `200` |
| `offset` | `0` | zero-based |

Each item contains: `sync_id`, `started_at`, `finished_at`, `total_files`, `new_files`, `removed_files`, `failed_files`, `sync_status`.

Status values: `syncing`, `out of sync`, `up to date`, `failed: <reason>`.

At most one in-progress `syncing` row exists per connector.

#### Example response

```json
{
  "total": 3,
  "limit": 50,
  "offset": 0,
  "items": [
    {
      "sync_id": 3,
      "started_at": "2025-01-15T14:32:00Z",
      "finished_at": "2025-01-15T14:32:10Z",
      "total_files": 42,
      "new_files": 0,
      "removed_files": 0,
      "failed_files": 0,

>       "sync_status": "up to date"

Reviewer: ```suggestion
      "id": 3,
      "started_at": "2025-01-15T14:32:00Z",
      "finished_at": "2025-01-15T14:32:10Z",
      "total_files": 42,
      "new_files": 0,
      "removed_files": 0,
      "failed_files": 0,
      "status": "up to date",
      "error": ""
```

let's remove redundant sync prefix
and we need to add error field here similar to the one available at connector level.
also now I m thinking the status at connector level should not be used as is here.
It is at per sync level, hence thinking started, completed & failed would make more sense here, wdyt?
cc @manju956

Author: Yes, makes sense to keep it simple. 
Error message (if any) on the latest sync will be the last_sync_error on GET connectors

Reviewer 2: makes sense to me.

Author: Done

Reviewer: > Error message (if any) on the latest sync will be the last_sync_error on GET connectors

Yes, we can update both connectors and sync table on every tick respectively

# Digitize — Data Source Connector Processing Proposal

> **Scope:** Internal `digitize` behavior after catalog sends connector payloads. Catalog-side concerns such as key management, connector CRUD, deployment wiring and TLS provisioning remain out of scope and are treated as infrastructure-level prerequisites.

---

## 1. Preconditions

Before any `digitize` connector endpoint is called:

- Catalog has already validated the remote connector configuration.
- Catalog sends secret material in plaintext via API calls:
  - `ssh`: `private_key`
  - `s3`: `secret_access_key`
- `digitize` encrypts those secret fields at rest using `/run/secrets/connector_encryption_key` before persisting them.
- `/run/secrets/connector_api_token` and `/run/secrets/connector_encryption_key` are mounted before pod start.

> - The `file_checksum_registry` table and the `FileChecksumRegistry` ORM model already exist in the codebase (merged in the de-duplication PR). This table is **user-submitted documents only** — connector code must never read from or write to it. Connector dedup is handled exclusively via `connector_file_membership` (see §4.4).

Reviewer: this is not required, we can remove all mention of file_checksum_registry table

| `private_key_ciphertext` | `string (base64)` | ✅ | AES-256-GCM encrypted Ed25519 private key PEM. Encrypted under the per-connector DEK. |
| `encrypted_dek` | `string (base64)` | ✅ | AES-256-GCM encrypted 32-byte DEK. Encrypted under the pod's `digitize_KEK` secret mount. |

**`connection_details` — `s3`:**

| Field | Type | Required | Description |
| --- | --- | --- | --- |
| `region` | `string` | ✅ | AWS region where the bucket resides (e.g. `"us-east-1"`). |
| `bucket_name` | `string` | ✅ | Exact name of the S3 bucket containing the files (e.g. `"my-rag-documents"`). |
| `access_key_id` | `string` | ✅ | IAM Access Key ID used for API authentication. Stored in plaintext (not a secret). |
| `secret_access_key_ciphertext` | `string (base64)` | ✅ | AES-256-GCM encrypted IAM Secret Access Key. Encrypted under the per-connector DEK. |
| `encrypted_dek` | `string (base64)` | ✅ | AES-256-GCM encrypted 32-byte DEK. Encrypted under the pod's `digitize_KEK` secret mount. |

**Response:**

| Status | Body | Meaning |
| --- | --- | --- |
| `202 Accepted` | `{ "connector_id": "c7f3a2d1-..." }` | Connector row created; worker thread started asynchronously via `BackgroundTask`. |

> | `409 Conflict` | Error detail | A connector with this `connector_id` already exists. Catalog should use `PUT` to update. |

Reviewer 3: This line contradicts line no 860:
`| `upsert_active_connector(config: dict)` | `active_connectors` | `INSERT … ON CONFLICT (id) DO UPDATE` — idempotent re-push from catalog. `connection_details` is stored as-is from `config["connection_details"]`. |`

Author: Fixed

       # if remaining > 0: another connector still holds this content — leave doc intact

4. DELETE active_connectors ROW
   delete_active_connector(connector_id)
   ← ON DELETE CASCADE removes any remaining connector_file_membership rows
     for this connector automatically (guards against any rows missed in step 3)
   ← file_checksum_registry is NOT cascaded — shared content survives

6. CLEANUP STAGING DIRECTORIES (best-effort)
   for staging_dir in glob(f"{settings.staging_dir}/connector-{connector_id}-*/"):
       shutil.rmtree(staging_dir, ignore_errors=True)
   ← Removes any per-tick staging directories that were left behind by
     an in-progress or failed tick. Uses ignore_errors=True so a locked
     file does not abort the handler.
```

**Error handling policy for the document-deletion loop (step 3):**

| HTTP status from `DELETE /v1/documents/{doc_id}` | Action |
| --- | --- |
| `200 OK` / `204 No Content` | Success — proceed to next entry |
| `404 Not Found` | Document already absent — treat as success; proceed |
| `5xx` / network error | Log warning with `doc_id` and `remote_path`; continue loop (best-effort cleanup) |

> Because step 5's `ON DELETE CASCADE` removes the registry unconditionally, a document-deletion failure means the document may remain in the VDB/storage but its registry row is gone. This is a known trade-off: a hard delete of the connector cannot be held hostage by a flaky document-deletion call. Operators can find and clean up orphaned documents via the `GET /v1/documents` listing filtered by `connector_id`.

Reviewer 3: you are mentioning a step 5 but step 5 is missing above.

Author: Fixed

Three new tables are added following the existing `IF NOT EXISTS` / idempotent DDL pattern already used in [`init_schema.sql`](../../services/digitize/db/scripts/init_schema.sql). All three tables also need corresponding SQLAlchemy ORM model classes added to [`db/models.py`](../../services/digitize/db/models.py), matching the `Job` / `Document` pattern there, so `Base.metadata.create_all()` on startup creates them automatically.

### 5.1 `active_connectors`

> **Security note:** There are **no plaintext secret columns**. All credential material (`private_key_ciphertext`, `secret_access_key_ciphertext`, `encrypted_dek`) is stored as AES-256-GCM ciphertext inside `connection_details`. A Postgres breach alone exposes nothing without the KEK.

The `connection_details` JSONB column stores all connector-type-specific fields — both non-sensitive config and encrypted credential blobs. The exact keys differ by `type`:

| `type` | Keys stored in `connection_details` |
| --- | --- |
| `ssh_sftp` | `port`, `username`, `remote_path`, `private_key_ciphertext`, `encrypted_dek` |
| `s3` | `region`, `bucket_name`, `access_key_id`, `secret_access_key_ciphertext`, `encrypted_dek` |

```sql

> CREATE TABLE IF NOT EXISTS active_connectors (

Reviewer 3: Shouldn't this contain last_sync_error as well?

Author: Not exactly, it is stored in connector_sync_history and can be retrieved with a query

     │
     │  AES-256-GCM decrypt(connection_details["private_key_ciphertext"], DEK)
     ▼
privkey_pem  (Ed25519 PEM string, in-memory only)
     │
     │  paramiko.Ed25519Key.from_private_key(io.StringIO(privkey_pem))
     │  privkey_pem string overwritten immediately after key load
     ▼
paramiko SSHClient  →  open SFTP session
     │
     ▼  close() — DEK not retained
```

**Key zeroization:** After calling `from_private_key()`, the `privkey_pem` variable is overwritten with `"\x00" * len(privkey_pem)` and set to `None`. The DEK is not retained between ticks — re-derived fresh on every tick to minimise in-memory exposure.

**KEK load timing (Decision D-6 — recommended):** The KEK is read fresh inside `SFTPScanner.connect()` on every tick. This keeps KEK residence time in memory minimal and allows future KEK rotation without a pod restart.

### 7.2 File Scanning — Streaming SHA-256, Filtered by Extension

File type filtering is applied at scan time using `allowed_extensions` from the connector config. This avoids SFTP transfers for files the ingest pipeline would reject. The extension list is populated by catalog (Decision D-8).

> `scan()` performs four phases:

Reviewer 3: nit: it's actually 5 steps below.

Author: Updated, PTAL

    │               continue  (counted as files_deduped)                   │
    │                                                                      │
    │           ← first time this content is seen globally: must ingest    │
    │           needs_ingest.append(file)                                  │
    │                                                                     │
    │  6. Store computed checksums for all diff-list files in the DB      │
    │     before download begins, so that a crash mid-batch does not      │
    │     leave the registry in a permanently inconsistent state.          │
    │                                                                     │
    │     NOTE: only files in needs_ingest reach this step — deduped      │
    │     files already have a registry row from the connector that        │
    │     originally ingested them.                                        │
    │                                                                     │
    │     For each file in needs_ingest:                                  │
    │         upsert_pending_checksum(id, file.checksum)                  │

>     │           ← lightweight write; doc_id is NULL until ingest succeeds │

Reviewer 3: How is this possible if line no 728 says:
` doc_id        TEXT        NOT NULL UNIQUE    `

Author: SHA is computed over stream during file walk. 
Used in memory for lookup and diff computation. 
Batch ingestion is done. 
Only after successful ingest, the checksums are stored and table populated.

**Manual sync on POST:** After `insert_active_connector` writes the row and `connector_worker_manager.start_worker()` starts the background thread, the handler enqueues an **immediate manual tick** by calling `worker.trigger_now()` (see §9.2). This sets a `threading.Event` on the worker so that the sleep at the bottom of the tick loop is interrupted and the next tick starts at once rather than waiting `sync_interval_seconds`.

**Request body (`application/json`):**

Fields common to **all** connector types are at the top level. All connector-specific credential and connection fields are nested inside `connection_details`.

**Example — `ssh_sftp`:**

```json
{
  "connector_id":          "c7f3a2d1-...",

>   "type":                  "ssh_sftp",

Reviewer: we can simply keep it as ssh.
sftp is an underlying detail on how are we doing this.

Author: Done

  "host":                  "sftp.example.com",

>   "allowed_extensions":    [".pdf", ".docx", ".xlsx"],

Reviewer: remove xlsx

Author: Done

>   "sync_interval_seconds": 300,

Reviewer: for this release let's configure this as env and use it for now.
let's not expose via post payload

Author: So it will not be configurable on UI also?

Reviewer: yes

Author: Done

  "connection_details": {

>     "port":                   22,

Reviewer: this is not configurable.
can be removed now

Author: Done

    "username":               "sync_user",
    "remote_path":            "/exports/reports",
    "private_key_ciphertext": "<base64-AES-256-GCM ciphertext of Ed25519 PEM>",

>     "encrypted_dek":          "<base64-AES-256-GCM ciphertext of 32-byte DEK under pod KEK>"

Reviewer: we can remove this.
and remove all logic related to kek and dek at the proposal level
seems like we would be receiving the private key as plain text, let's keep only that.

Author: Done

  }
}
```

**Example — `s3`:**

```json
{
  "connector_id":          "a1b2c3d4-...",
  "type":                  "s3",
  "host":                  "s3.amazonaws.com",
  "allowed_extensions":    [".pdf", ".docx", ".xlsx"],
  "sync_interval_seconds": 300,

>   "connection_details": {

Reviewer: need to add prefix and delimiter

Author: Done

**`connection_details` — `s3`:**

| Field | Type | Required | Description |
| --- | --- | --- | --- |
| `region` | `string` | ✅ | AWS region where the bucket resides (e.g. `"us-east-1"`). |
| `bucket_name` | `string` | ✅ | Exact name of the S3 bucket containing the files (e.g. `"my-rag-documents"`). |
| `access_key_id` | `string` | ✅ | IAM Access Key ID used for API authentication. Stored in plaintext (not a secret). |
| `secret_access_key_ciphertext` | `string (base64)` | ✅ | AES-256-GCM encrypted IAM Secret Access Key. Encrypted under the per-connector DEK. |
| `encrypted_dek` | `string (base64)` | ✅ | AES-256-GCM encrypted 32-byte DEK. Encrypted under the pod's `digitize_KEK` secret mount. |

**Response:**

| Status | Body | Meaning |
| --- | --- | --- |
| `202 Accepted` | `{ "connector_id": "c7f3a2d1-..." }` | Connector row created; worker thread started asynchronously via `BackgroundTask`. |
| `409 Conflict` | Error detail | A connector with this `connector_id` already exists. Catalog should use `PUT` to update. |
| `401 Unauthorized` | Error detail | Missing or invalid bearer token. |

---

### B. `PUT /v1/connectors/{connector_id}` — Update an existing connector

> **Purpose:** Called by catalog when the configuration of an existing connector changes — for example, credentials are rotated, the remote path or bucket changes, the sync interval is adjusted, or `allowed_extensions` is updated. Digitize stops the existing sync worker, applies the updated config via `upsert_active_connector`, restarts a fresh worker with the new settings, and **immediately triggers a manual sync tick** on the new worker so that any changes take effect at once.

Reviewer: as per our discussion, we won't trigger a manual sync on this, we would throw an error if there is an existing sync happening on the source. we would allow only if the connector is in synced/error state

Author: Done, no more manual triggers. Not hot-update for PUT call either, we will wait for current sync to finish and at start of each tick, fetch from DB

# Digitize — Data Source Connector Processing Proposal

> > **Scope:** Internal `digitize` behavior after catalog sends connector payloads. Catalog-side concerns such as key generation, connector CRUD, deployment wiring, TLS provisioning, and bearer-token enforcement remain out of scope and are treated as infrastructure-level prerequisites.

Reviewer: ```suggestion
> **Scope:** Internal `digitize` behavior after catalog sends connector payloads. Catalog-side concerns such as key management, connector CRUD, deployment wiring and TLS provisioning remain out of scope and are treated as infrastructure-level prerequisites.
```

Author: Done

---

## 1. Preconditions

Before any `digitize` connector endpoint is called:

- Catalog has already validated the remote connector configuration.

> - Catalog sends secret material in plaintext over TLS:

Reviewer: TLS is till agent or catalog backend.
From there to digitize is just internal podman network without TLS.

Author: Done, reworded it

  - `ssh`: `private_key`
  - `s3`: `secret_access_key`
- `digitize` encrypts those secret fields at rest using `/run/secrets/connector_encryption_key` before persisting them.
- `/run/secrets/connector_api_token` and `/run/secrets/connector_encryption_key` are mounted before pod start.

> - TLS and bearer-token enforcement are already provided by infrastructure.

Reviewer: this line can be removed

Author: Removed

  → write sync status + history
```

### 2.3 Main Components

- `active_connectors`: current connector configuration and top-level sync state
- `file_checksum_registry`: global content registry keyed by `sha256`
- `connector_file_membership`: which connector currently references which hash
- `connector_sync_history`: one row per worker tick
- `ConnectorWorkerManager`: owns worker thread lifecycle
- `ConnectorSyncWorker`: executes periodic sync logic
- scanner implementations: transport-specific remote access for SFTP and S3

---

## 3. API Contract

All endpoints are bearer-token protected and served over TLS.

### 3.1 `POST /v1/connectors`

Creates a connector, stores encrypted credentials, persists config, and schedules worker startup asynchronously. The worker runs its first tick immediately after the thread starts.

#### Request body

Common fields:

| Field | Type | Required | Notes |
| --- | --- | --- | --- |
| `connector_id` | `string (UUID)` | ✅ | Stable catalog ID |
| `type` | `string` | ✅ | `ssh` or `s3` |
| `host` | `string` | ✅ | SFTP host or S3 endpoint |
| `allowed_extensions` | `array[string]` | ✅ | Non-matching files are ignored |

> | `sync_interval_seconds` | `integer` | ✅ | Poll interval |

Reviewer: let's remove this and configure via env in this release.

Author: Done

`connection_details` for `s3`:

| Field | Type | Required |
| --- | --- | --- |
| `bucket_name` | `string` | ✅ |
| `prefix` | `string` | ❌ |
| `delimiter` | `string` | ❌ |
| `access_key_id` | `string` | ✅ |
| `secret_access_key` | `string` | ✅ |

#### Example payloads

```json
{
  "connector_id": "c7f3a2d1-...",
  "type": "ssh",
  "host": "sftp.example.com",
  "allowed_extensions": [".pdf", ".docx"],
  "sync_interval_seconds": 300,
  "connection_details": {
    "username": "sync_user",
    "remote_path": "/exports/reports",
    "private_key": "-----BEGIN OPENSSH PRIVATE KEY-----\n..."
  }
}
```

```json
{
  "connector_id": "a1b2c3d4-...",
  "type": "s3",
  "host": "s3.amazonaws.com",
  "allowed_extensions": [".pdf", ".docx"],
  "sync_interval_seconds": 300,
  "connection_details": {
    "bucket_name": "my-rag-documents",
    "prefix": "reports/",
    "delimiter": "/",
    "access_key_id": "AKIAIOSFODNN7EXAMPLE",
    "secret_access_key": "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
  }
}
```

#### Response codes

| Status | Meaning |
| --- | --- |
| `202 Accepted` | Connector created; worker start scheduled in a background task |
| `409 Conflict` | Connector already exists |

> | `401 Unauthorized` | Missing or invalid token |

Reviewer: when is this possible?

Author: Removed it. Both Boto and Paramiko rely on throwing exceptions in case credentials fails

### 3.2 `PUT /v1/connectors/{connector_id}`

> 

Reviewer: Mention there would be validation against the last sync status, if it is in progress, request need to be rejected

Reviewer: I think we need to revisit the rejection part.
Maybe we can allow the db update and next sync onwards it would point to the new details.

Reviewer: Do you see any problems with that?
@sats-23 @manju956

Author: Yes, I thought that was the plan, keep the db update async, because either way we are polling from DB before starting every fresh sync tick

Updates an existing connector's config in the database. The running worker is not restarted — it reads the latest config from the DB before entering the next tick.

Rules:

- All fields are optional.
- Omitted fields remain unchanged.
- `type` cannot change.
- `connection_details` is merged by key, not replaced wholesale.
- If credentials are included, they are re-encrypted before storage.

Example partial update:

```json
{
  "sync_interval_seconds": 600,
  "connection_details": {
    "remote_path": "/exports/v2/reports",
    "private_key": "-----BEGIN OPENSSH PRIVATE KEY-----\n..."
  }
}
```

Response codes:

| Status | Meaning |
| --- | --- |
| `200 OK` | Connector updated; worker picks up changes on next tick |
| `404 Not Found` | Connector does not exist |

> | `401 Unauthorized` | Missing or invalid token |

Reviewer: here too

Author: Removed

### 3.3 `DELETE /v1/connectors/{connector_id}`

Removes a connector and its runtime state.

Delete flow:

1. Stop the worker.
2. Snapshot the connector's known hashes.
3. Remove membership rows hash by hash.
4. Delete documents only when the remaining reference count reaches zero.
5. Delete the `active_connectors` row.
6. Best-effort cleanup of staging directories.

#### Delete sequence diagram

```text
DELETE /v1/connectors/{connector_id}
  → stop worker
  → list connector hashes
  → for each sha256:
       remove membership in transaction
       count remaining refs
       if refs == 0:
         DELETE /v1/documents/{doc_id}
  → delete connector row
  → cleanup staging dirs
```

Document deletion is best-effort:

- `200`, `204`, `404` from `DELETE /v1/documents/{doc_id}` are treated as success.
- `5xx` or network failures are logged and cleanup continues.

Response codes:

| Status | Meaning |
| --- | --- |
| `204 No Content` | Connector removed |
| `404 Not Found` | Connector does not exist |
| `401 Unauthorized` | Missing or invalid token |

> ### 3.4 `GET /v1/connectors`

Reviewer: Please add json payload sample for this and remaining GET calls

Author: Done

Lists active connectors with non-secret configuration and current sync state.

Returned fields include:

> - connector identity and config

Reviewer: what configs are you planning to include?

Author: Planned was connector_id, type, host, allowed_extensions, sync_interval_seconds

- `sync_status`, `last_sync_at`, `last_sync_error`, `attached_at`

> - non-secret `connection_details`

Reviewer: connection_details can be part of `/v1/connectors/{connector_id}` no need in GET all call

Author: Removed

- `N files failed to ingest`
- `N orphan deletes failed`
- `failed: <reason>`

At most one in-progress `syncing` row exists per connector.

---

## 4. Data Model

**Modified file:** `services/digitize/db/scripts/init_schema.sql`

### 4.1 Table Relationships

```text
active_connectors
  └─< connector_file_membership >─┐
                                  └─ file_checksum_registry ─> documents

active_connectors
  └─< connector_sync_history
```

### 4.2 `active_connectors`

Stores connector config, encrypted credential blobs, and top-level sync state.

```sql
CREATE TABLE IF NOT EXISTS active_connectors (
    id                      TEXT        PRIMARY KEY,
    type                    TEXT        NOT NULL,
    host                    TEXT        NOT NULL,
    connection_details      JSONB       NOT NULL DEFAULT '{}',
    allowed_extensions      JSONB       NOT NULL DEFAULT '[]',
    sync_interval_seconds   INTEGER     NOT NULL DEFAULT 300,
    attached_at             TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    last_sync_at            TIMESTAMPTZ,
    sync_status             TEXT        NOT NULL DEFAULT 'idle',
    CONSTRAINT chk_connector_type CHECK (type IN ('ssh', 's3'))
);
```

> ### 4.3 `file_checksum_registry`

Reviewer: Will connector or sync worker try to add an entry into this table?

Author: Nope, the batch job will take care of adding entries to file_checksum_registry. 
The Sync Worker thread will only work with connector_file_membership.

| `ingested_this_tick` | `(sha256, doc_id)` pairs | After each successful ingest |

DELETE flow:

1. `stop_event.set()` — sleeping thread wakes immediately; ticking thread stops at next boundary.
2. `thread.join(timeout=30 s)`.
3. If still alive after timeout: log warning and proceed with cleanup best-effort.

Key invariants:

| Invariant | Mechanism |
| --- | --- |
| Interval never interrupts a running tick | Timer checked only after `_run_tick()` returns |
| DELETE cooperatively interrupts at next boundary | `_cancel_if_requested()` after each phase |
| No concurrent ticks per connector | Single-threaded loop + tick guard |
| Monitor skips intentionally stopped threads | `if stop_event.is_set(): skip` |

### 10.6 Lifespan Recovery

On FastAPI startup: load `active_connectors`, recreate workers, start the monitor. No catalog re-push needed.

---

## 11. Implementation Plan — Digitize Connector PRs

Each PR is independently testable — no PR leaves things in a broken or untestable state.

---

### PR 1 — DB Schema + ORM Models + Settings

**Files touched:** `init_schema.sql`, `db/models.py`, `config/settings.py`

**What's built:**

> - 4 new tables: `active_connectors`, `file_checksum_registry`, `connector_file_membership`, `connector_sync_history`

Author: Some of these details will update once de-dup PR is merged

    sync_status             TEXT        NOT NULL DEFAULT 'idle',
    CONSTRAINT chk_connector_type CHECK (type IN ('ssh', 's3'))
);
```

> **Note:** `sync_interval_seconds` is stored per-connector for future extensibility but is not accepted via the API today. On `POST`, it is populated from the `CONNECTOR_SYNC_INTERVAL_SECONDS` environment variable (default `300`). The worker reads the value from the DB before each tick.

### 4.3 `file_checksum_registry`

Global content registry keyed by `sha256`. One row per unique document body.

```sql
CREATE TABLE IF NOT EXISTS file_checksum_registry (
    sha256        TEXT PRIMARY KEY,
    doc_id        TEXT NOT NULL UNIQUE REFERENCES documents(doc_id) ON DELETE CASCADE
);
```

### 4.4 `connector_file_membership`

Maps connectors to hashes they currently own.

```sql
CREATE TABLE IF NOT EXISTS connector_file_membership (
    connector_id  TEXT NOT NULL,
    sha256        TEXT NOT NULL,
    PRIMARY KEY (connector_id, sha256),
    FOREIGN KEY (connector_id) REFERENCES active_connectors(id) ON DELETE CASCADE,
    FOREIGN KEY (sha256) REFERENCES file_checksum_registry(sha256)
);
```

Deduplication and deletion rules:

> - Existing `sha256` in `file_checksum_registry` => reuse document and add membership only.

Author: @dharaneeshvrd @manju956 
I feel like we should keep the insert operations to connector_file_membership as part of the main thread (instead of individual worker thread) and executed along with insert operations to file_checksum_registry (in same transaction).

This comes from PoV that a file membership can be owned by a manual ingest by user as well (which we never store). So if a connector is deleted, we risk removing the files that the user intended to manually ingest as well. 

Minor change, but I feel its good to keep the pass the owner/connectorID through the create_job loop so the main thread can execute this DB operation 
Needed because:
1. Stores membership for files ingested manually as "user" or "manual" to prevent being deleted when connector is dropped. 
2. Prevents race condition and PK violation when multiple threads are adding/removing entries from membership table

Reviewer: Take care of below things:

What is the field you are thinking to add to job properties?
What would be the value of it in case of user triggered job?
We should not show this property on GET calls of job.

Author: Sure, all seem to be implementation specifics. I shall not mention in proposal

    "access_key_id": "AKIAIOSFODNN7EXAMPLE",
    "secret_access_key": "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
  }
}
```

#### Response codes

| Status | Meaning |
| --- | --- |
| `202 Accepted` | Connector created; worker start scheduled in a background task |
| `409 Conflict` | Connector already exists |

### 3.2 `PUT /v1/connectors/{connector_id}`

Updates an existing connector's config in the database. The running worker is not restarted — it reads the latest config from the DB before entering the next tick.

Rules:

- All fields are optional.
- Omitted fields remain unchanged.
- `type` cannot change.
- `connection_details` is merged by key, not replaced wholesale.
- If credentials are included, they are re-encrypted before storage.
- `sync_interval_seconds` cannot be set via this endpoint; change the env variable and redeploy.

Example partial update:

```json
{
  "connection_details": {
    "remote_path": "/exports/v2/reports",
    "private_key": "-----BEGIN OPENSSH PRIVATE KEY-----\n..."
  }
}
```

Response codes:

| Status | Meaning |
| --- | --- |
| `200 OK` | Connector updated; worker picks up changes on next tick |
| `404 Not Found` | Connector does not exist |

> ### 3.3 `DELETE /v1/connectors/{connector_id}`

Reviewer: Please mention that DELETE would be allowed only if there are no active sync happening.
Since stopping active sync requires, force stopping the jobs which is not there currently, mention that as a stretch, if that's accomplished, DELETE can be allowed even when a sync is happening.

Author: Done

5. Delete the `active_connectors` row.
6. Best-effort cleanup of staging directories.

#### Delete sequence diagram

```text
DELETE /v1/connectors/{connector_id}
  → stop worker
  → list connector hashes
  → for each sha256:
       remove membership in transaction
       count remaining refs
       if refs == 0:
         DELETE /v1/documents/{doc_id}
  → delete connector row
  → cleanup staging dirs
```

Document deletion is best-effort:

- `200`, `204`, `404` from `DELETE /v1/documents/{doc_id}` are treated as success.
- `5xx` or network failures are logged and cleanup continues.

Response codes:

| Status | Meaning |
| --- | --- |
| `204 No Content` | Connector removed |
| `404 Not Found` | Connector does not exist |

### 3.4 `GET /v1/connectors`

Lists active connectors with non-secret configuration and current sync state.

Returned fields include:

- connector identity and config
- `sync_status`, `last_sync_at`, `last_sync_error`, `attached_at`

#### Example response

```json
[
  {
    "connector_id": "c7f3a2d1-4e5b-4c6d-8f9a-0b1c2d3e4f5a",
    "type": "ssh",
    "host": "sftp.example.com",
    "allowed_extensions": [".pdf", ".docx"],

>     "sync_interval_seconds": 300,

Reviewer: no need.

Author: Done

    "attached_at": "2025-01-10T08:00:00Z",
    "last_sync_at": "2025-01-15T14:32:10Z",
    "sync_status": "idle",

>     "last_sync_error": null,

Reviewer: we discussed to attach last sync's details as well right?
to show the file stats.

Reviewer: nevermind.
sry that's on GET of a specific connector.

  → scan source — yields (remote_path, checksum) per file
  → compare checksums with known_checksums from connector_file_membership
  → skip files whose checksum is already in connector_file_membership for this connector
  → ingest only new content
  → create-job is called with connector_id — main thread inserts into
     connector_file_membership (keyed by checksum, carrying connector_id + doc_id)
  → store source checksum in document metadata
  → remove orphaned content by ref count
  → write sync status + history

User-submitted files (no connector_id)
  → create-job flow detects absence of connector_id
  → dedup check against file_checksum_registry (sha256 PK) only
  → on success: insert row into file_checksum_registry (sha256, doc_id)
  → connector_file_membership is never written for user-submitted docs
```

### 2.3 Main Components

- `active_connectors`: current connector configuration and top-level sync state
- `file_checksum_registry`: **user-submitted documents only** — keyed by `sha256`, stores a content fingerprint and the `doc_id`. **Already implemented** — table and ORM model exist. Connector code must never write to this table.

> - `connector_file_membership`: **connector-sourced documents only** — keyed by `checksum` (PK), carries the list of connector IDs that reference this file and the `doc_id` for deletion. User-submitted docs are never written here.

Reviewer 2: So, we will have SHA for user submitted files and md5 for connector sourced files?

Author: No, everything will be MD5, I shall make that change

    )
```

**User-submitted flow:** when a user submits a file via the API or the Digitize UI:

1. Compute `sha256` of the file content.
2. Check `file_checksum_registry` — if already present, return `409 / already_exists`.
3. On successful ingest, insert `(sha256, doc_id)` into `file_checksum_registry`.

**Connector code must not write here.** The connector path uses `connector_file_membership` exclusively.

### 4.4 `connector_file_membership`

**Connector-sourced documents only.** This table is the sole dedup and reference-counting store for all content ingested via connectors. User-submitted code must never write to this table.

> The schema uses `checksum` as the primary key. The `connector_ids` column is a JSON array of all connector IDs that currently reference this file (cross-connector shared-content tracking). `doc_id` is stored here — not in `file_checksum_registry` — so that deletion can proceed without a registry join.

Reviewer 2: why does `connector_ids` have to be JSON array not a simple array?

Author: It just chosen for cosmetics reason for some operations like removing element by value.
Here's a comparison:

| Operation | TEXT[] Equivalent | JSONB Equivalent |
| :--- | :--- | :--- |
| Remove element by value | `array_remove(col, val)` | `col - val` |
| Containment check | `col @> ARRAY[val]` | `col @> jsonb_build_array(val)` |
| Append element | `col \|\| ARRAY[val]` | `col \|\| jsonb_build_array(val)` |
| Length check | `array_length(col, 1)` | `jsonb_array_length(col)` |

The performance difference is negligible. 

What do you think? Should we switch to TEXT[] or keep JSONB

| `delete_active_connector()` | delete connector |
| `update_connector_sync_status()` | write top-level sync state |
| `merge_connection_details()` | key-level JSON merge for PUT |
| `lookup_content_by_checksum(checksum)` | dedup lookup — returns `doc_id` or `None` |
| `insert_checksum_registry(checksum, source, doc_id)` | register newly ingested content |
| `delete_checksum_registry(checksum)` | defensive cleanup |
| `list_connector_checksums(connector_id)` | current checksums owned by this connector |
| `insert_connector_membership(connector_id, checksum)` | add ownership row |
| `delete_connector_membership_atomic(connector_id, checksum)` | remove ownership, return remaining ref count |
| `insert_sync_history()` | create tick row |
| `update_sync_history()` | finalize tick row |
| `update_sync_history_files_syncing()` | live progress updates |
| `list_sync_history()` | paginated history query |

> | `set_document_metadata(doc_id, metadata)` | write `source_checksum` + S3 key into `documents.metadata` |

Author: Why S3 key?

Reviewer 2: this is basically the s3 bucket path to trace where the document is sourced from. I think we dont need, since we already have connector_id->checksum mapping in `connector_membership` table.

Subclass responsibilities:

> - remote listing — yields `(remote_path, checksum)` pairs

Author: Why path?
